consensus/ibft/validators/bls.go

```diff
@@ -0,0 +1,303 @@
+		x, y := (*v)[indx], (*target)[indx]
+
+		if !x.Equals(&y) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// Copy returns a clone of ECDSAValidatorSet
+func (v *BLSValidatorSet) Copy() ValidatorSet {
+	vals := make([]BLSValidator, v.Len())
+
+	for i := range vals {
+		vals[i].Address = (*v)[i].Address
+		vals[i].BLSPubKey = make([]byte, len((*v)[i].BLSPubKey))
+		copy(vals[i].BLSPubKey, (*v)[i].BLSPubKey)
+	}
+
+	clone := BLSValidatorSet(vals)
+
+	return &clone
+}
+
+//
+func (v *BLSValidatorSet) GetAddress(index int) types.Address {
+	if index < v.Len() {
+		return (*v)[index].Address
+	}
+
+	return types.ZeroAddress
+}
+
+// Index returns the index of the passed in address in the validator set.
+// Returns -1 if not found
+func (v *BLSValidatorSet) Index(target types.Address) int {
+	for index, addr := range *v {
+		if addr.Address == target {
+			return index
+		}
+	}
+
+	return -1
+}
+
+// Includes checks if the address is in the validator set
+func (v *BLSValidatorSet) Includes(addr types.Address) bool {
+	return v.Index(addr) != -1
+}
+
+// CalcProposer calculates the address of the next proposer, from the validator set
+func (v *BLSValidatorSet) CalcProposer(round uint64, lastProposer types.Address) types.Address {
+	var seed uint64
+
+	if lastProposer == types.ZeroAddress {
+		seed = round
+	} else {
+		offset := 0
+		if indx := v.Index(lastProposer); indx != -1 {
+			offset = indx
+		}
+
+		seed = uint64(offset) + round + 1
+	}
+
+	pick := seed % uint64(v.Len())
+
+	return (*v)[pick].Address
+}
+
+// MaxFaultyNodes returns the maximum number of allowed faulty nodes (F), based on the current validator set
+func (v *BLSValidatorSet) MaxFaultyNodes() int {
+	return CalcMaxFaultyNodes(v)
+}
+
+// Add adds a new address to the validator set
+func (v *BLSValidatorSet) Add(addr types.Address) {
+	// TODO: doesn't support right now
+}
+
+// Del removes an address from the validator set
+func (v *BLSValidatorSet) Del(target types.Address) {
+	// TODO: doesn't support right now
+}
+
+// Merge adds set of validators
+func (v *BLSValidatorSet) Merge(newRawSet ValidatorSet) error {
+	newSet, ok := newRawSet.(*BLSValidatorSet)
+	if !ok {
+		return fmt.Errorf("can't merge with BLSValidatorSet and %T", newRawSet)
+	}
+
+	for _, newVal := range *newSet {
+		if !v.Includes(newVal.Address) {
+			*v = append(*v, newVal)
+		}
+	}
+
+	return nil
+}
+
+func (v *BLSValidatorSet) MarshalRLPWith(ar *fastrlp.Arena) *fastrlp.Value {
+	if len(*v) == 0 {
+		return ar.NewNullArray()
+	}
+
+	valSet := ar.NewArray()
+
+	for _, val := range *v {
+		val := val
+		valSet.Set(val.MarshalRLPWith(ar))
+	}
+
+	return valSet
+}
+
+func (v *BLSValidatorSet) UnmarshalRLPFrom(p *fastrlp.Parser, vv *fastrlp.Value) error {
+	vals, err := vv.GetElems()
+	if err != nil {
+		return fmt.Errorf("mismatch of RLP type for ECDSAValidatorSet, expected list but found %s", vv.Type())
+	}
+
+	valSet := make([]BLSValidator, len(vals))
+
+	for index := range vals {
+		vals := vals
```

Comment:
Why the identity assignment ?

Reply:
Resolving because this code has been removed already